subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/ApiGroovyCompiler.java

```diff
@@ -73,8 +74,8 @@ public ApiGroovyCompiler(Compiler<JavaCompileSpec> javaCompiler) {
 
     private static abstract class IncrementalCompilationCustomizer extends CompilationCustomizer {
         static IncrementalCompilationCustomizer fromSpec(GroovyJavaJointCompileSpec spec) {
-            if (spec.getCompilationMappingFile() != null) {
-                return new TrackingClassGenerationCompilationCustomizer(spec.getCompilationMappingFile());
+            if (spec.getCompilationMappingFile() != null && !spec.getSourceRoots().isEmpty()) {
```

Comment:
We should not be configuring a compilation mapping file if the source roots cannot be inferred. Let's not use the magic `isEmpty` here and disable incremental compilation altogether.

Reply:
Done.